### PR TITLE
[BottomSheetBehavior] Check for shouldSkipHalfExpandedWhenDragging during nested scrolling

### PR DESCRIPTION
**Pull request by OxygenCobalt**
Make `BottomSheetBehavior` check the experimental `shouldSkipHalfExpandedWhenDragging` flag before setting the state to `HALF_EXPANDED`.

The code changes the following behaviors in `onStopNestedScroll`:
1. In the case of `dy > 0`, the state will be set to `STATE_EXPANDED` instead of `STATE_HALF_EXPANDED` when the flag is enabled.
2. In the case of `dy == 0`, the state will be set to `STATE_EXPANDED` when above the half-expanded offset, and `STATE_COLLAPSED` when below it when the flag is enabled.
3. In the state of `dy < 0`, the state will be set to `STATE_COLLAPSED` instead of `STATE_HALF_EXPANDED` when the flag is enabled.

### CODE REVIEW

File: lib/java/com/google/android/material/bottomsheet/BottomSheetBehavior.java [hunk 1]
```diff
@@ -796,10 +800,14 @@ public void onStopNestedScroll(
             }
           }
         } else {
-          if (Math.abs(currentTop - halfExpandedOffset) < Math.abs(currentTop - collapsedOffset)) {
-            targetState = STATE_HALF_EXPANDED;
-          } else {
+          if (shouldSkipHalfExpandedStateWhenDragging()) {
```
Comment (drchen) [inline lib/java/com/google/android/material/bottomsheet/BottomSheetBehavior.java:803]: if (shouldSkipHalfExpandedStateWhenDragging() || Math.abs(currentTop - halfExpandedOffset) >= Math.abs(currentTop - collapsedOffset)) {

to avoid the below nested if

File: lib/java/com/google/android/material/bottomsheet/BottomSheetBehavior.java [hunk 2]
```diff
@@ -809,10 +817,14 @@ public void onStopNestedScroll(
       } else {
         // Settle to nearest height.
         int currentTop = child.getTop();
-        if (Math.abs(currentTop - halfExpandedOffset) < Math.abs(currentTop - collapsedOffset)) {
-          targetState = STATE_HALF_EXPANDED;
-        } else {
+        if (shouldSkipHalfExpandedStateWhenDragging()) {
```
Comment (drchen) [inline lib/java/com/google/android/material/bottomsheet/BottomSheetBehavior.java:820]: ditto

File: lib/java/com/google/android/material/bottomsheet/BottomSheetBehavior.java (hunk 1 above)
Comment (OxygenCobalt) [reply lib/java/com/google/android/material/bottomsheet/BottomSheetBehavior.java]: Completed in 7564954

File: lib/java/com/google/android/material/bottomsheet/BottomSheetBehavior.java (hunk 2 above)
Comment (OxygenCobalt) [reply lib/java/com/google/android/material/bottomsheet/BottomSheetBehavior.java]: Completed in 7564954.

**Comment by OxygenCobalt** [comment]
Okay, simplified those if statements @drchen.

**Review by drchen — APPROVED** [review]

**Comment by OxygenCobalt** [comment]
Completely forgot about this PR. Any updates on this too? @drchen

**Comment by drchen** [comment]
We found some problems during the internal review. 

I need to take more time to understand the relevant logic and figure out what's the cause. : )

**Comment by OxygenCobalt** [comment]
Could you elaborate on the issue @drchen? I'm using this fix in my app, so it might affect me. I could try looking into it further.

**Comment by drchen** [comment]
My teammate patched and tested the PR and said:

"""
...the scroll behavior is almost completely broken for this case -- it doesn't expand upwards when you try to drag it up. it looks like this is because it only behaves properly when shouldExpandOnUpwardsDrag() is also set to true. wdyt about removing that flag? i can't imagine a scenario where we'd want it set to false. also, that shouldExpandOnUpwardsDrag() only takes effect when used with shouldSkipHalfExpandedStateWhenDragging(), which is pretty unexpected. i think we should delete that flag and just add another shouldSkipHalfExpandedStateWhenDragging() check next to the fitToContents offset check in onViewReleased().
"""

Haven't really got time to check it yet.

**Comment by OxygenCobalt** [comment]
Oh yeah, I already know about that bug @drchen. It was present before my fix, and I'm really not sure what the point behind it was.

I could go look into it and try to remove that function. Could I append it to this PR or should I go and make a new one?

**Comment by OxygenCobalt** [comment]
Any updates here @drchen? Should I make another PR to remove `shouldExpandOnUpwardsDrag` so this can get unblocked?

**Comment by drchen** [comment]
Hi the change is pretty difficult to land the change due to all the corner cases we found during running internal tests.

I'll need a less busy timespan to go back working on this. Sorry for the long delay.

**Comment by OxygenCobalt** [comment]
Sorry for taking awhile to respond @hellosagar. I don't really mind if you copy this changeset into your own vendored bottom sheet.
